ProboCI/probo-asset-receiver
# Feature/asset size migration

This builds off 
https://github.com/ProboCI/probo-asset-receiver/pull/6
and 
https://github.com/ProboCI/probo-asset-receiver/pull/8

Creates migration scripts that can be run from the bin directory. These accomplish three things:
1. Rekey the levelDB data to make it possible to see a list of buckets without seeing the files and file versions in those buckets as well as making it possible to see asssets or asset versions by bucket.
2. Capture the raw and zipped file size in bytes or all previously uploaded files and store that metadata in levelDB along with the filename.
3. The files are re-ciphered using values unique to each bucket and asset and saved back to disk.

### To Test and Deploy
1. Run `./migrations/probo-leveldb-key-update -c /etc/probo/asset-receiver.yaml`
2. Check the levelDb keys and ensure that match the new format (see below)
3. Ensure that you have the directtory to capture the re-ciphered files `mkdir -p /var/lib/probo/asset-receiver-files--reciphered`
4. Run `./migrations/probo-asset-fix-and-upload -c /etc/probo/asset-receiver.yaml`
5. Check levelDb values for assets and ensure that they have file size and name metadata.
6. Swap the file directories in the asset-reciever.yaml file or physically move the new reciphered files to the asset-reciever's file directory.
7. Use the download endpoint to ensure you can decrypt one of the newly reciphered files.
8. Ensure you can upload and download new files.

OLD FORMAT levelDB Keys
```
asset!!54ab466e3dc00000
bucket!!0f1b1044-4f0a-4aa3-9040-55355c932d93
bucket!!e92a3075-a88f-4771-a169-23c001e22ccd!!asset!!wordpress.sql.gz
bucket!!e92a3075-a88f-4771-a169-23c001e22ccd!!mydb.sql.gz!!54a8cbec09400000
bucket-token!!e92a3075-a88f-4771-a169-23c001e22ccd!!4fc92e22-ba04-46b5-a0ae-e0ed3a0f759e
token!!03df74b8-64ed-428d-8c01-bee0c1a66415

```
NEW FORMAT levelDB Keys
```
asset!!54ab466e3dc00000
bucket!!0f1b1044-4f0a-4aa3-9040-55355c932d93
bucket-asset!!e92a3075-a88f-4771-a169-23c001e22ccd!!wordpress.sql.gz
bucket-asset-version!!e92a3075-a88f-4771-a169-23c001e22ccd!!mydb.sql.gz!!54a8cbec09400000
bucket-token!!e92a3075-a88f-4771-a169-23c001e22ccd!!4fc92e22-ba04-46b5-a0ae-e0ed3a0f759e
token!!03df74b8-64ed-428d-8c01-bee0c1a66415
```
